Rfc proposal (#1)
FlixCoder/rfcs: text/0000-iterator-try_all-try_any.md

- Feature Name: iter_try_all_any
- Start Date: 2022-02-15
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

> Add `try_any` and `try_all` functions to the `Iterator` trait, which work similar to `any` and `all`, but accept `Result` and `Option` as closure return type.

Reviewer: ```suggestion
Add the associated methods `try_any` and `try_all` to the `Iterator` trait, which work similar to `any` and `all`, but return `Result` and `Option` from the given closure.
```

# Motivation
[motivation]: #motivation

In some cases where one would like to use `any`/`all`, the actual condition to check is fallible. For example, let's say we have a simple key-value-store like a `HashMap`. The keys are IDs to the objects. We further have a list of IDs we want to check some condition on. We would like to use `all` on an iterator over the list to ensure all objects match our condition. However, access to elements is fallable and returns `Option`s. In other cases, the checks might involve more complex validation on the objects. The validation method could involve access to external parts and fail. Therefore, it returns a `Result<bool, Error>`.

> The implementor needs to return a `bool` for every element in `all`. They could use `.try_fold(true, |state, item| Ok(state && check(item)?))`, but this is less readable and does not short-circuit as `all` does. If they would do `.all(|item| check(item).unwrap_or(false))`, the error would be discarded and lost/ignored. To fully comply with expectations, there is only two possibilites (I see):

Reviewer: ```suggestion
The implementor needs to return a `bool` for every element in `all`. They could use `.try_fold(true, |state, item| Ok(state && check(item)?))`, but this is less readable and does not short-circuit as `all` does. If they would do `.all(|item| check(item).unwrap_or(false))`, the error would be discarded and lost/ignored. To fully achieve `try_all` behavior:
```

- a `for` loop is required to iterate over the elements and stop early on errors or invalid items
- collecting the elements in the iterator to a collected `Result`, then calling `all` on the inner collection

Both solutions are not very elegant. The `try_all` method would work as `.try_all(check)`, returning `Result<bool, Error>` as well. It should behave similar to `all`, except working with `Try<bool>` output, just as `try_fold` behaves in comparison to `fold`.

Everything applies similarly to `try_any`.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

> These additions of `try_all` and `try_any` to the core library of Rust allows the use of fallible validators in iterators.

Reviewer: ```suggestion
These additions of `try_all` and `try_any` to the core library of Rust allow the use of fallible validators in iterators.
```

When every element in an iterator needs to be validated, it could be done like this:
```rust
fn file_is_empty(path: &str) -> std::io::Result<bool> {
	Ok(std::fs::read_to_string(path)?.is_empty())
}

fn main() {
	let files = ["fileA.txt", "fileB.txt"];
	let all_exist_but_empty = files.iter().try_all(file_is_empty).expect("file wasn't found");
	let any_exists_but_empty = files.iter().try_any(file_is_empty).expect("file wasn't found");
}
```

It works similarly with `Option`s, e.g. the mentioned key-value-store:
```rust
fn is_valid(object: &MyObject) -> bool {
	*object != MyObject::default()
}

fn main() {
	let mut kv = HashMap::new();
	kv.insert("id-1", MyObject::default());
	
	let object_list = ["id-1", "id-5"];
	let all_valid = object_list.iter().try_all(|id| kv.get(id).map(is_valid)).expect("id wasn't found");
}
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

> The addition `try_all` behaves to `all` like `try_fold` behaves to `fold`. The similar applies to `try_any`. They both should react as their simple counter-part: short-circuit on elements that already determine the overall result. However two things are different:

Reviewer: ```suggestion
The addition `try_all` behaves to `all` like `try_fold` behaves to `fold`. Similarly it applies to `try_any`. They both should react as their simple counter-part: short-circuit on elements that already determine the overall result. However two things are different:
```
